pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -477,7 +477,20 @@ func applyHTTPRouteDestination(
 	out.Action = &route.Route_Route{Route: action}
 
 	if in.Rewrite != nil {
-		action.PrefixRewrite = in.Rewrite.GetUri()
+		action.ClusterSpecifier = &route.RouteAction_Cluster{
+			Cluster: in.Name,
+		}
+		uri := in.Rewrite.GetUri()
+		if fullURI, isFullPathRewrite := cutPrefix(uri, "*fullreplace*"); isFullPathRewrite && model.UseGatewaySemantics(vs) {
```

Comment:
Do we use this pattern in any other part of the code ? Is this something we want to do - introduce a  "*action*" prefix ? Anyone else using this pattern ? 

I personally don't mind having a structured replacement syntax - we support %VAR% and a lot of other things, but *action* seems unusual. We could have something like

%regexRewrite "/.+" "fullURI"% ( just an idea - and probably not a good one...)

Reply:
I think we should have no customization in core generator

Reply:
The best thing would be to change the public api to support it I guess.
But since it is still only an experimental feature in gw-api, I think this can be a good start.


I am happy to change to a public api change if we all think it is better.
I assume we can also add a hidden field to `istio.HttpRewrite{}` but I am wondering why would we ? I think it would be better to also add this support for non gateway-api users but I might be missing something.

@howardjohn curios to know your thoughts

Reply:
My only concern with adding it to the public API is that the field in GW api is experimental. So we would be adding a stable field to support experimental API which is concerning if it changes in some way.

Reply:
I don't mind having an expression as rewrite value - headers rewrite also use expressions, I haven't checked yet if path rewrite does support the %ENVOY_MAGIC% but I assume it does. So rewrite value is already not really a string, but an evaluated expression.

And if value is already an expression - I don't think we need to add another field in the API that is also an expression.

My concern is introducing a strange syntax to the expression language we support.  We have %ENVOY_KEY%, in some parts we use CEL - *fullreplace* as a prefix doesn't sound right. 

How about %ISTIO_FULLREPLACE% as a prefix ? Note that this will also allow users to do full replace using the beta API if they really want ( it would be an Istio-specific extension, like all other % variables). But I would go one step further and use %ISTIO_REGEX% as prefix and expose the real envoy API. It will obviously be experimental initially.

Reply:
summarizing my understanding from our offline chat.
I will go with %ISTIO_FULLREPLACE(<replacement-path>)% 

and should come up with a better proposal of how(and if) we want to insert substitutions to istio (similar to what envoy supports [here](https://github.com/envoyproxy/envoy/blob/main/source/common/formatter/substitution_formatter.cc)

Thoughts? @costinm @howardjohn

Reply:
A couple of thoughts as a passer-by:

* There seems to be an issue with using Istio protos as an intermediate representation. A more direct or more expressive IR between gateway APIs and xDS would avoid these magic "fullreplace" strings.
* Envoy optimizes prefix regexes by not actually performing regex and doing a simple suffix test.  Header values are not interpreted as UTF-8 inside Envoy so a lot of complexity around UTF matching in RE2 is meaningless.

Reply:
Agree with first point - I think in ~6 months ( or after HttpRoute usage is larger than VirtualService ) we could do some cleanup and use HttpRoute as model. We may also see how the Envoy implementation of Gateway API looks like and maybe align. But for now - I think it's good enough and we have other things to worry about.

 It's good that they optimize it - but from API perspective it's still a regex and confusing for users with all the alternate syntax variations and complexity.